flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectOfInputTypeStrategy.java

```diff
@@ -0,0 +1,162 @@
+ *   <li>{@code OBJECT_OF('com.example.User', 'name', 'Alice')} - single field
+ *   <li>{@code OBJECT_OF('com.example.User', 'name', 'Alice', 'age', 30)} - multiple fields
+ * </ul>
+ *
+ * @see org.apache.flink.table.functions.BuiltInFunctionDefinitions#OBJECT_OF
+ * @see ObjectOfTypeStrategy
+ */
+public class ObjectOfInputTypeStrategy implements InputTypeStrategy {
+
+    private static final ArgumentCount AT_LEAST_ONE_ODD =
+            new ArgumentCount() {
+                @Override
+                public boolean isValidCount(int count) {
+                    return count % 2 == 1;
+                }
+
+                @Override
+                public Optional<Integer> getMinCount() {
+                    return Optional.of(1);
+                }
+
+                @Override
+                public Optional<Integer> getMaxCount() {
+                    return Optional.empty();
+                }
+            };
+
+    private static void validateFieldName(
+            final CallContext callContext,
+            final int idx,
+            final LogicalType logicalType,
+            final Set<String> filedNames) {
+        final int keyIndex = (idx + 1);
+        if (!logicalType.is(LogicalTypeFamily.CHARACTER_STRING)) {
+            throw new ValidationException(
+                    "The field key at position "
+                            + keyIndex
+                            + " must be a STRING/VARCHAR type, but was "
+                            + logicalType.asSummaryString()
+                            + ".");
+        }
+        final Optional<String> filedName = callContext.getArgumentValue(idx, String.class);
+        filedName.ifPresent(
+                name -> {
+                    if (!filedNames.add(name)) {
+                        throw new ValidationException(
+                                "The field name '"
+                                        + name
+                                        + "' at position "
+                                        + keyIndex
+                                        + " is repeated.");
+                    }
+                });
+    }
+
+    @Override
+    public ArgumentCount getArgumentCount() {
+        return AT_LEAST_ONE_ODD;
+    }
+
+    @Override
+    public Optional<List<DataType>> inferInputTypes(
+            final CallContext callContext, final boolean throwOnFailure) {
+        final List<DataType> argumentDataTypes = callContext.getArgumentDataTypes();
+        final LogicalType classTypeArgument = argumentDataTypes.get(0).getLogicalType();
+
+        if (!classTypeArgument.is(LogicalTypeFamily.CHARACTER_STRING)) {
+            throw new ValidationException(
+                    "The first argument must be a STRING/VARCHAR type representing the class name.");
+        }
+
+        final Set<String> fieldNames = new HashSet<>();
+        for (int i = 1; i < argumentDataTypes.size(); i++) {
+            if (i % 2 != 0) {
```

Comment:
is there a reason to iterate on each index if we have such condition?
Or can we just go through odd elements only?